tests/h/auth/util_test.py

```diff
@@ -171,6 +171,35 @@ def test_translate_annotation_principals(p_in, p_out):
     assert set(result) == set(p_out)
 
 
+class TestClientAuthority(object):
+
+    @pytest.mark.parametrize('principals', [
+        ['foo', 'bar', 'baz'],
+        ['authority', 'foo'],
+        [],
+        ['authority:'],
+        [' authority:biz.biz', 'foo'],
+        ['authority :biz.biz', 'foo'],
+    ])
+    def test_it_returns_None_if_no_authority_principal_match(self, principals):
+        class FakePrincipalsRequest(object):
+            """With a real request, you cannot set the ``effective_principals`` property"""
+            effective_principals = principals
```

Comment:
This is fine, but an alternative would be to create a spec based on `pyramid_request` using eg. `request = mock.create_autospec(type(pyramid_request), instance=True, spec_set=True)`. This would protect from eg. typos in the attribute name.

Reply:
Actually, would you mind changing this @lyzadanger ? I think there's a couple of important things to understand and get right here, for the integrity of our tests, and I particularly want the two of you to understand these as it goes beyond just these two tests to any future tests that you write or review. This sort of thing comes up when Pyramid auth stuff is tested, but also in other causes such as when tests need routes to be set up. I understand that this part of Pyramid isn't the clearest.

1. We should never be using a mock or fake request object in the tests, not even with autospec. I know that there are probably existing tests that do use both mock and fake requests, but those tests are wrong.

   Autospec won't work correctly for the request object in all cases.

2. No "real" request objects are ever used in the tests either.

3. What the tests always use, or should always use, is a [DummyRequest](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html#pyramid.testing.DummyRequest) object. This isn't a real request object, it's a simulated one that Pyramid provides specifically for the purpose of making it easy to set request properties (such as `effective_principals`) in tests.

4. More specifically, the tests should always use [our `pyramid_request` fixture](https://github.com/hypothesis/h/blob/3e8ee139174669c1dee2474021e9a098a72092b8/tests/h/conftest.py#L236-L249) which is an h-specific `DummyRequest` instance.

5. It's not true that you can't set `effective_principals` on `pyramid_request`. You can, but you have to use [`testing.security_policy`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/config.html#pyramid.config.Configurator.testing_securitypolicy) which is the thing that Pyramid provides for doing exactly this and related things in tests.

6. Reading on this:

  * [The `pyramid.testing` module](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html) (Pyramid's unit test helpers, including `DummyRequest` and others).
  * More [Testing Helper APIs in `pyramid.config`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/config.html#pyramid.config.Configurator.testing_add_renderer) (includes `testing_securitypolicy` and others)
  * [Unit, Integration, and Functional Testing chapter in the Pyramid docs](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/testing.html)
  * [Our global test fixtures](https://github.com/hypothesis/h/blob/master/tests/h/conftest.py), the `pyramid_request` and `pyramid_config` ones are particularly relevant here.
  * Grep our existing tests for `testing_securitypolicy` and `pyramid_config` for examples

There correct way to write these two tests it like this. Try it - if you break the code under test, the tests will fail.

```python
    @pytest.mark.parametrize(
        "principals",
        [
            ["foo", "bar", "baz"],
            ["authority", "foo"],
            [],
            ["authority:"],
            [" authority:biz.biz", "foo"],
            ["authority :biz.biz", "foo"],
        ],
    )
    def test_it_returns_None_if_no_authority_principal_match(
        self, principals, pyramid_request, pyramid_config
    ):
        pyramid_config.testing_securitypolicy("seanh", groupids=principals)

        assert util.client_authority(pyramid_request) is None

    @pytest.mark.parametrize(
        "principals,authority",
        [
            (["foo", "bar", "baz", "authority:felicitous.com"], "felicitous.com"),
            (["authority:somebody.likes.me", "foo"], "somebody.likes.me"),
        ],
    )
    def test_it_returns_authority_if_authority_principal_matchpyramid_requesi(
        self, principals, authority, pyramid_request, pyramid_config
    ):
        pyramid_config.testing_securitypolicy("seanh", groupids=principals)

        assert util.client_authority(pyramid_request) == authority
```

Reply:
Depending on how this `client_authority()` helper is going to be called, could it possibly be called like `client_authority(request.effective_principals)` rather than `client_authority(request)`? That would make testing easier as you just pass a list of strings into it, rather than a request object. On the other hand maybe we don't want the calling code to have to know that the client authority is based on effective principals in particular?

Reply:
Thanks for the super-detailed response @seanh. I have encountered `testing_securitypolicy` in past but it completely skipped my mind here.